### PR TITLE
TDSLoader: Use path and manager for all Loaders

### PR DESCRIPTION
**Pull request by cnspaha**
If a path and a manager is given I can't understand why the path is only used for textures and the manager only used for the .3ds file...
If a loading manager is available, there is no reason to not use it in all loaders. Especially when I pass an LoadingManager to an Loader instance/class I would assume that it is used for all Loaders, not for just some.

If for whatever reason you need two different loading managers then another function for setting the second one should be created.

@tentone 

@Mugen87 , got another one here, all loaders behave different... :/

### CODE REVIEW

**Comment by tentone** [comment]
@cnspaha Thanks a lot for these changes, this happened because initially i wrote this code for my threejs based engine (https://github.com/tentone/nunuStudio) that has a  different resource management system, missed that during the refactoring!

Thanks a lot, it really does not make any sense to have two load managers inside the same loader :+1:

**Comment by Mugen87** [comment]
> got another one here, all loaders behave different... :/

Yeah, the loaders were developed by many contributors. It's not surprising to see such differences.

**Comment by Mugen87** [comment]
BTW: We should not merge this until we have a proper solution for #12614

**Comment by cnspaha** [comment]
@Mugen87 
should we try to make the path logic like in https://github.com/mrdoob/three.js/pull/12614#discussion_r150025998 ?

**Comment by Mugen87** [comment]
Yes, i think `ColladaLoader` and `GLTFLoader` are now good prototypes for all other loaders.

**Comment by Itee** [comment]
I'm not totaly agree with that ! They should implement worker management and Logger like ObjLoader2 too ! In that way they will be perfect :stuck_out_tongue_winking_eye:

**Comment by tentone** [comment]
@Itee Totally agree some loader print out a lot of unnecessary debug information, i think that we should at least have a debug flag in them with false as default value.

**Comment by Itee** [comment]
Just a little note about manager: #12617 (see MrDoob comment about manager extend)

**Comment by Mugen87** [comment]
> They should implement worker management and Logger like ObjLoader2 too !

I think there are enough use cases where the overhead of those components are unnecessary. I prefer to keep the implementation of the loaders as simple as possible. So i would not vote for this suggestion.

**Comment by cnspaha** [comment]
@Mugen87 
> Yes, i think ColladaLoader and GLTFLoader are now good prototypes for all other loaders.

Made the path logic fit the ColladaLoader/GLTFLoader one.

File: examples/js/loaders/TDSLoader.js
Comment (Mugen87) [inline examples/js/loaders/TDSLoader.js]: I think we don't need this line.

**Review by Mugen87 — CHANGES_REQUESTED** [review]

File: examples/js/loaders/TDSLoader.js
Comment (tentone) [reply examples/js/loaders/TDSLoader.js]: That line sets the new path (deferred from URL) when there is no path defined

File: examples/js/loaders/TDSLoader.js
Comment (tentone) [inline examples/js/loaders/TDSLoader.js]: I would change this for a if condition and only dedefine the path "load.setPath" if there was no path defined.

Also,  path is never undefined it starts defined as "" to avoid error on string concatenation.

Comment (Mugen87) [reply examples/js/loaders/TDSLoader.js]: @cnspaha The implementation of `TDSLoader` varies from `GLTFLoader` and `ColladaLoader`. So we can't do just a copy/paste of the code unless we change parts of `TDSLoader`. But i don't think this is necessary.

@tentone Since you are the original author, i'll leave it up to you to guide @cnspaha through the PR 😊

Comment (cnspaha) [reply examples/js/loaders/TDSLoader.js]: @Mugen87 
yeah, you are right.

@tentone 
since you know the good/functionality of the TDSLoader better than me, would you mind to create a PR with the correct code, regarding the issues from this PR?

Comment (tentone) [reply examples/js/loaders/TDSLoader.js]: @cnspaha 

I think we can keep this PR, the file seems to be good i would only change the load method to this

```
	load: function ( url, onLoad, onProgress, onError ) {
	
		var scope = this;

		if ( this.path === undefined )	{

			this.path = THREE.Loader.prototype.extractUrlBase( url );

		}
		
		var loader = new THREE.FileLoader( this.manager );

		loader.setResponseType( 'arraybuffer' );
		
		loader.setPath( this.path );

		loader.load( url, function ( data ) {

			onLoad( scope.parse( data ) );

		}, onProgress, onError );

	},
```

Comment (cnspaha) [reply examples/js/loaders/TDSLoader.js]: @tentone 
but as you said yourself ```this.path``` is always defined with an empty string, so
```this.path = THREE.Loader.prototype.extractUrlBase( url );``` is never reached? So the whole if branch is useless?

Comment (tentone) [reply examples/js/loaders/TDSLoader.js]: Oh my bad, youre right missed that, it shoud be this instead

if ( this.path === "" ) {
	this.path = THREE.Loader.prototype.extractUrlBase( url );
}

Comment (cnspaha) [reply examples/js/loaders/TDSLoader.js]: yeah, sounds reasonable to me ;)
will wait for @Mugen87 s opinion about your load method and if hes also fine with it I will commit it ;)

Comment (Mugen87) [reply examples/js/loaders/TDSLoader.js]: Go for it 👍

File: examples/js/loaders/TDSLoader.js
Comment (Mugen87) [reply examples/js/loaders/TDSLoader.js]: Um, i've missed the different approach in this loader.

File: examples/js/loaders/TDSLoader.js
Comment (cnspaha) [reply examples/js/loaders/TDSLoader.js]: done :)

**Comment by cnspaha** [comment]
@Mugen87 @tentone 
fixed the requested changes, would you mind to remove your "Changes requested" and approve it, if there are no other issues? ;)

File: examples/js/loaders/TDSLoader.js
Comment (Mugen87) [reply examples/js/loaders/TDSLoader.js]: Please consider this comment: https://github.com/mrdoob/three.js/pull/12614#discussion_r150025998

Wouldn't be the mentioned case also problematic for `TDSLoader`? 

\cc @donmccurdy

**Comment by donmccurdy** [comment]
@cnspaha there's an open suggestion here still: https://github.com/mrdoob/three.js/pull/12615#discussion_r150491208

The gist is that calling `load()` should not modify `this.path`, but just keep track of a local path variable for the duration of that model's parsing.

**Comment by donmccurdy** [comment]
> > They should implement worker management and Logger like ObjLoader2 too !
> 
> I think there are enough use cases where the overhead of those components is unnecessary. I prefer to keep the implementation of the loaders as simple as possible. So i would not vote for this suggestion.

This probably needs another issue to discuss, but I would also like to have a simple way to silence or intercept logging (ideally without pulling in all of LoaderSupport?).

IMO worker management should remain a case-by-case thing — my hunch is that loading glTF in a worker would actually introduce _more_ work on the main thread, because the glTF format is already heavily optimized for runtime parsing and what the worker serializes back may not be as optimal. Specific subtasks of the glTF loader (e.g. DRACO decompression) could certainly benefit from using a worker though.

**Comment by cnspaha** [comment]
> @cnspaha there's an open suggestion here still: #12615 (comment)

haha yeah, that is the problem when working on multiple branches on multiple loaders :D 
some things are getting overseen/forgotten... 

added commit with local variable

**Comment by Mugen87** [comment]
That does not solve the problem. The entire statement needs to be deleted otherwise you overwrite `this.path`. The actual problem with `TDSLoader` is that is does not pass the local `path`  variable of `.load()` as a parameter to the `.parse()` method. That's the way `GLTFLoader` and `ColladaLoader` works.

**Comment by cnspaha** [comment]
The "problem" I currently seeing here is also that if we would pass the "path" to the parse function (which I would also recommend) that we would have to pass it down 4 function calls before it is needed again.  So readFile, readMeshData and readMaterialEntry would each need a "path" parameter and would just have to pass through it to the next function till readMap is reached. (see image)
image

Sure it would be possible to just pass through the parameter but I don't really like this idea. 
using `this.path` like it is currently is of course also out of the question.

What do you think about adding a `this.currentPath` ? 
Per default it would be an empty string or undefined and instead of using a local `path` variable which we would have to pass through several functions (without doing anything there) we could later just access the `this.currentPath`. Of course the variable would be set back on each `.load`.  So it would be like a single-use (only for one file) path.
Do you understand what I am trying to say?

**Comment by Mugen87** [comment]
I think the best solution is to re-write `TDSLoader` according to `GLTFLoader`. So the only public methods are `.load()`, `.parse()`, `.setCrossOrigin()` and `.setPath()`. All other code (which is related to parsing) should be encapsulated in separate functions. These functions could be nested inside the `.parse()` method (like `ColladaLoader`) or inside a top-level closure (like `GLTFLoader`). In this way, it's much easier to handle the mentioned scoping issue. And we would have a clean public interface of the loader.

**Comment by cnspaha** [comment]
@Mugen87 
Yeah, I also already thought about a complete rewrite of the TDSLoader.  Just thought maybe we find another solution, since I am totaly stuck in work here and won't have time, at least not in the near future, to do a complete rewrite of it.  
But a rewrite would be the best

**Comment by Mugen87** [comment]
I suggest to merge this PR since it adds the `manager` instance to all internal loaders and improves the `path` handling. The mentioned refactoring can be done in a separate PR.

**Review by Mugen87 — APPROVED** [review]

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**
